clippy_lints/src/mutable_debug_assertion.rs

```diff
@@ -0,0 +1,83 @@
+use crate::utils::{is_expn_of, match_qpath, span_note_and_lint};
+use rustc::hir::{Expr, ExprKind, Mutability, UnOp};
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use rustc::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for function calls that mutate an object in `debug_assert!`, `debug_assert_eq!` and
+    /// `debug_assert_ne!` macros.
+    ///
+    /// **Why is this bad?** In release builds the code won't be executed and therefore can alter
+    /// the execution flow.
+    ///
+    /// **Known problems:** None
+    ///
+    /// **Example:**
+    /// ```rust,ignore
+    /// debug_assert_eq!(vec![3].pop(), Some(3));
+    /// // or
+    /// let mut s = String::from("foo");
+    /// debug_assert_eq!(s.remove(0), 'f');
+    /// ```
+    pub DEBUG_ASSERT_WITH_MUT_CALL,
+    correctness,
```

Comment:
I don't feel well by making this an hard error. Any better category?

Reply:
Why not? Because you feel like the lint has too many false positives? In that case, you can put it in `nursery` until it is ready.

If you think that the lint doesn't have FPs, `correctness` is the most fitting group IMO.